pkg/common/jwt/validator.go

```diff
@@ -0,0 +1,77 @@
+package jwt
+
+import (
+	"context"
+	"crypto"
+	"fmt"
+
+	"github.com/HewlettPackard/galadriel/pkg/common/keymanager"
+	"github.com/golang-jwt/jwt/v4"
+)
+
+// Validator validates JWT tokens using a public key.
+type Validator interface {
+	// ValidateToken ValidateJWT validates a JWT and returns the claims.
+	ValidateToken(context.Context, string) (*jwt.RegisteredClaims, error)
+}
+
+type ValidatorConfig struct {
+	// KeyManager is the key manager used to get the public key for validating the JWT.
+	KeyManager       keymanager.KeyManager
+	ExpectedAudience []string
+}
+
+type DefaultJWTValidator struct {
+	keyManager       keymanager.KeyManager
+	expectedAudience []string
+}
+
+func NewDefaultJWTValidator(c *ValidatorConfig) *DefaultJWTValidator {
+	return &DefaultJWTValidator{
+		keyManager:       c.KeyManager,
+		expectedAudience: c.ExpectedAudience,
+	}
+}
+
+func (v *DefaultJWTValidator) ValidateToken(ctx context.Context, token string) (*jwt.RegisteredClaims, error) {
+	claims := &jwt.RegisteredClaims{}
+	_, err := jwt.ParseWithClaims(token, claims, func(token *jwt.Token) (interface{}, error) {
+		return v.getPublicKey(ctx, token)
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse and validate token: %w", err)
+	}
+
+	// Validate the claims
+	err = claims.Valid()
+	if err != nil {
+		return nil, err
+	}
+
+	validAudience := true
+	for _, aud := range v.expectedAudience {
+		ok := claims.VerifyAudience(aud, true)
+		if !ok {
+			validAudience = false
+			break
+		}
+
+	}
+
+	if !validAudience {
+		return nil, jwt.NewValidationError("token audience is invalid", jwt.ValidationErrorAudience)
+	}
+
+	return claims, nil
+}
```

Comment:
I think that encapsulating audience verification would make the code simpler. Something similar to this:
```suggestion
func (v *DefaultJWTValidator) ValidateToken(ctx context.Context, token string) (*jwt.RegisteredClaims, error) {
	claims := &jwt.RegisteredClaims{}
	_, err := jwt.ParseWithClaims(token, claims, func(token *jwt.Token) (interface{}, error) {
		return v.getPublicKey(ctx, token)
	})
	if err != nil {
		return nil, fmt.Errorf("failed to parse and validate token: %w", err)
	}

	// Validate the claims
	err = claims.Valid()
	if err != nil {
		return nil, err
	}
	
	if !v.validAudience(claims) {
		return nil, jwt.NewValidationError("token audience is invalid", jwt.ValidationErrorAudience)
	}

	return claims, nil
}

func (v *DefaultJWTValidator) validAudience(claims *jwt.RegisteredClaims) bool {
	for _, aud := range v.expectedAudience {
		ok := claims.VerifyAudience(aud, true)
		if !ok {
			return false
		}
	}
	return true
}
```